auth.go

```diff
@@ -0,0 +1,127 @@
+package mysql
+
+import "bytes"
+
+const mysqlClearPassword = "mysql_clear_password"
+const mysqlNativePassword = "mysql_native_password"
+const mysqlOldPassword = "mysql_old_password"
+const defaultAuthPluginName = mysqlNativePassword
+
+var authPluginFactories map[string]func(*Config) AuthPlugin
+
+func init() {
+	authPluginFactories = make(map[string]func(*Config) AuthPlugin)
+	authPluginFactories[mysqlClearPassword] = func(cfg *Config) AuthPlugin {
+		return &clearTextPlugin{cfg}
+	}
+	authPluginFactories[mysqlNativePassword] = func(cfg *Config) AuthPlugin {
+		return &nativePasswordPlugin{cfg}
+	}
+	authPluginFactories[mysqlOldPassword] = func(cfg *Config) AuthPlugin {
+		return &oldPasswordPlugin{cfg}
+	}
+}
+
+// RegisterAuthPlugin registers an authentication plugin to be used during
+// negotiation with the server. If a plugin with the given name already exists,
+// it will be overwritten.
+func RegisterAuthPlugin(name string, factory func(*Config) AuthPlugin) {
+	authPluginFactories[name] = factory
+}
+
+// AuthPlugin handles authenticating a user.
+type AuthPlugin interface {
+	// Next takes a server's challenge and returns
+	// the bytes to send back or an error.
+	Next(challenge []byte) ([]byte, error)
```

Comment:
This is called next because it might be called more than once. All the current plugins, clear, native, and old all are single step. Next is called once and they are complete. However, not all plugins will be this way, so calling it Auth and having it called multiple times is weird (to me). Precedent was pulled from Java's driver, which calls it Next as well. However, I'm certainly not tied to that name.